public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts

```diff
@@ -0,0 +1,197 @@
+import { getSituation, Situation } from './situation';
+
+// we use the `^` character as the cursor-marker in the string.
+function assertSituation(situation: string, expectedSituation: Situation | null) {
+  // first we find the cursor-position
+  const pos = situation.indexOf('^');
+  if (pos === -1) {
+    throw new Error('cursor missing');
+  }
+
+  // we remove the cursor-marker from the string
+  const text = situation.replace('^', '');
+
+  // sanity check, make sure no more cursor-markers remain
+  if (text.indexOf('^') !== -1) {
+    throw new Error('multiple cursors');
+  }
+
+  const result = getSituation(text, pos);
+
+  if (expectedSituation === null) {
+    expect(result).toStrictEqual(null);
+  } else {
+    expect(result).toMatchObject(expectedSituation);
+  }
+}
+
+describe('situation', () => {
+  it('handles things', () => {
+    assertSituation('^', {
+      type: 'EMPTY',
+    });
+
+    assertSituation('s^', {
+      type: 'AT_ROOT',
+    });
+
+    assertSituation('{level="info"} ^', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    // should not trigger AFTER_SELECTOR before the selector
+    assertSituation('^ {level="info"}', null);
+
+    // check for an error we had during the implementation
+    assertSituation('{level="info" ^', null);
+
+    assertSituation('{level="info"} | json ^', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    assertSituation('{level="info"} | json | ^', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: true,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    assertSituation('count_over_time({level="info"}^[10s])', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    // should not trigger AFTER_SELECTOR before the selector
+    assertSituation('count_over_time(^{level="info"}[10s])', null);
+
+    // should work even when the query is half-complete
+    assertSituation('count_over_time({level="info"}^)', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    /*
+    Currently failing, reason unknown
```

Comment:
@gabor @ivanahuckova Gonna need some help here.

The current code in `situation.ts` expects the following path for `sum(`: `[ERROR_NODE_NAME, 'VectorAggregationExpr']`, and the parser is returning:

```
[
    'ERROR_NODE_NAME',
    'LiteralExpr',
    'MetricExpr',
    'VectorAggregationExpr',
    'MetricExpr',
    'Expr',
    'LogQL'
]
```

I'm not sure if I need to update the path with this other one, or if there is a bug in the code.

Reply:
@matyax in general, since the hackathon-code is quite old, if lezer-logql produces a different tree currently, i would just update the "hackathon-code".

but please make sure to always have tests in `situation.test.ts`, the largest reason why there is a sepratate `situation.ts` and a `completions.ts` is that `situation.ts` is very easy/efficient to unit-test.

also, it made sense from a logical side (one deals with "find out what is the situation we are in", other deals with "what should we offer in this situation")